File: LeetCode/23. Merge k Sorted Lists/23.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        heap=[]
        for i,node in enumerate(lists):
            if node:
                heapq.heappush(heap,(node.val,i,node))

        dummy=tail=ListNode(0)
        while heap:
            val,i,node=heapq.heappop(heap)
            tail.next=node
            tail=tail.next
            node=node.next
            if node:
                heapq.heappush(heap,(node.val,i,node))
        return dummy.next

        

File: LeetCode/23. Merge k Sorted Lists/23.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<ListNode> queue = new PriorityQueue<>(Comparator.comparingInt(a -> a.val));
        for(ListNode node:lists){
            if(node!=null){
                queue.add(node);
            }
        }
        ListNode dummy=new ListNode(0);
        ListNode tail=dummy;
        while(!queue.isEmpty()){
            ListNode minNode=queue.poll();
            tail.next=minNode;
            tail=tail.next;
            if(minNode.next!=null){
                queue.add(minNode.next);
            }
        }
        return dummy.next;

    }
}